src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/StaffConfigManager.kt

```diff
@@ -45,131 +50,125 @@ class StaffConfigManager(
     }
 
-    if (createConfig) {
-      request.apply {
-        staffConfigRepository.save(
-          StaffConfiguration(
-            referenceDataRepository.getReferenceData(
-              ReferenceDataKey(ReferenceDataDomain.STAFF_STATUS, if (status.isPresent) status.get() else "ACTIVE"),
-            ),
-            capacity.orElse(prisonConfig.maximumCapacity),
-            allowAutoAllocation.orElse(true),
-            reactivateOn.orElse(null),
-            staffId,
-          ),
-        )
+    val staffRole: StaffRoleInfo? = getStaffRoleIfExists(prisonCode, staffId, request)
+    val roleAction =
+      request.staffRole.map {
+        when {
+          request.staffRole.get() == null -> Action.REMOVE
+          staffRole == null -> Action.CREATE
+          else -> Action.UPDATE
+        }
       }
-    } else if (!deactivateJobRoleAndConfig) {
-      staffConfig.patch(request)
+    if (roleAction == Action.CREATE && !request.staffRole.get()!!.isValidToCreate()) {
+      throw ResponseStatusException(HttpStatusCode.valueOf(400), "Insufficient parameters to create staff job role")
     }
 
-    if (removeActiveAllocations) {
+    if (request.deactivateActiveAllocations.orElse(false) || roleAction == Action.REMOVE) {
       deallocateAllocationsFor(prisonCode, staffId)
     }
 
-    if (createRole) {
-      val jobRoleRequest =
-        request.staffRole.get()!!.let {
-          StaffJobClassificationRequest(
-            position = it.position.get(),
-            scheduleType = it.scheduleType.get(),
-            hoursPerWeek = it.hoursPerWeek.get(),
-            fromDate =
-              it.fromDate.orElse(
-                LocalDate.now(),
-              ),
-            toDate = null,
-          )
-        }
-      policy.nomisUseRoleCode?.let {
-        nurApi.setStaffRole(prisonCode, staffId, it, jobRoleRequest)
-      } ?: setStaffRole(prisonCode, staffId, jobRoleRequest)
-    } else if (patchRole) {
-      policy.nomisUseRoleCode?.let { roleCode ->
-        staffRole!!.second!!.apply {
-          nurApi.setStaffRole(
-            prisonCode,
-            staffId,
-            roleCode,
-            request.staffRole.get()!!.let {
-              StaffJobClassificationRequest(
-                position = it.position.orElse(position),
-                scheduleType = it.scheduleType.orElse(scheduleType),
-                hoursPerWeek = it.hoursPerWeek.orElse(hoursPerWeek),
-                fromDate = it.fromDate.orElse(fromDate),
-                toDate = null,
-              )
-            },
-          )
-        }
-      } ?: run {
-        staffRole!!.first!!.apply {
-          request.staffRole.get()!!.let { jobRoleRequest ->
-            jobRoleRequest.position.ifPresent {
-              position =
-                referenceDataRepository.getReferenceData(ReferenceDataDomain.STAFF_POSITION of it)
+    when (roleAction) {
+      null -> return
+      Action.CREATE -> {
```

Comment:
Could maybe break these action implementations out into separate functions to make this function a bit easier to read but it's already a big improvement